[gosongying] iP

gosongying opened the pull request:
# Harper

> "Your mind is for having ideas, not holding them." - David Allen [(source)](https://dansilvestre.com/productivity-quotes)

Harper frees your mind of having to remember things you need to do. It is,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/gosongying/ip/releases/download/A-Jar/duke-1.0.jar).
2. launch it with the command `java -jar duke-1.0.jar`.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here is the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Duke duke = new Duke("data", "harper.txt");
        duke.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
chatbot can greet the user and exit

Commit:
chatbot can echo user's input and exit when "bye" is entered.

Commit:
chatbox now can store user's input and display back when requested.

Commit:
user can mark tasks as done or not done. Task class is craeted to keep track of the task status.

Commit:
Chatbot now can handle three types of tasks including ToDo, Deadline and Event.

Commit:
Semi-automated UI testing and corrects some typo

Commit:
Created several exception classes to handle invalid input.

Commit:
User can delete tasks in the list.

Commit:
Save and load data in and from the hard disk

When the chatbot starts, it loads data from the hard disk. A folder named data and a file named harper.txt are created if not exist. When the task list changes, data are restored into the file. When the content of the file does not follow the expected format, the chatbot will terminate.

Commit:
Merge branch 'branch-Level-7'

Commit:
Enhance date and time handling for task

Introduce the use of java.time.LocalDateTime to store date and time for Deadline and Event task. Display the date and time in a more user-friendly way: d MMM yyyy HH:mm (e.g., 20 Oct 2024 12:12).

Commit:
Merge branch 'branch-Level-8'

Commit:
Reorganise the code into more classes to deadl with different part of the chatbot

Ui is creadted to deadl with interactions with user. Storage is created to dead with interactions with the file. TaskList is created to contain the tasks. Parser is created to make sense the command.

Commit:
Merge branch 'A-MoreOOP'

Commit:
Reorganise different classes into corresponding packages

Package harper is the root package. Inside that, package commands contains all classes that handle different commands. Package exceptions contains all the exception classes. Package tasks contains three types of task. Package utils contains classes that handle different jobs.

Commit:
Merge branch 'A-Packages'

Commit:
Merge branch 'add-gradle-support'

Commit:
Set up gradle to build and run the program

mainClass is changed to "harper.Duke".

Commit:
Unit testing for AddCommand, Parser and Storage using JUnit

Commit:
Merge branch 'A-JUnit'

Commit:
Package the app as a JAR file

Commit:
Merge branch 'A-Jar'

Commit:
Add JavaDoc comment to all classes and methods

Commit:
Refactor some code to comply with coding standard

Commit:
Add find command

FindCommand class is created to represents that command,
Parser is updated to handle that command,
TaskList is updated to find matching tasks.

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/harper/utils/Ui.java

Xuyan0518 commented on src/main/java/harper/commands/Command.java:
```diff
@@ -0,0 +1,26 @@
+import harper.utils.Storage;
+
+public abstract class Command {
+    private boolean isExit;
```
Good job on naming the boolean variable!

Xuyan0518 commented on src/main/java/harper/Duke.java:
```diff
@@ -0,0 +1,53 @@
+package harper;
```
Good job! The package's names are all in small letters.

Xuyan0518 commented on src/main/java/harper/Duke.java:
```diff
@@ -0,0 +1,53 @@
+    private TaskList taskList;
+    private Ui ui;
+
+    public Duke(String folderName, String fileName) {
```
Nice! The variable names are in camel case.

Xuyan0518 commented on src/main/java/harper/utils/Parser.java:
```diff
@@ -0,0 +1,170 @@
+     * @param command commands.Command entered by user.
+     * @return Different type of commands.Command based on user's input.
+     */
+    public static Command parse(String command) {
```
Nice! The indentation is correct for if-else blocks.

Xuyan0518 commented on src/test/java/harper/utils/StorageTest.java:
```diff
@@ -0,0 +1,26 @@
+    public void load_correctFileFormat_success() {
+        Storage storage = new Storage("src/test", "test2.txt");
+
+        assertDoesNotThrow(() -> storage.load());
```
Well-done! The spacing within statements is correct!

Xuyan0518 commented on src/main/java/harper/commands/Command.java:
```diff
@@ -0,0 +1,26 @@
+     * @param taskList utilities.TaskList to be operated.
+     * @param ui utilities.Ui that handles input and output.
+     * @param storage utilities.Storage that handles operation with hard disk.
+     */
```
The format of Javadoc is correct!

Xuyan0518 reviewed (commented):
Good job! Java coding standard is strictly followed!

Commit:
Configure checkstyle and correct error

Add two config files checkstyle.xml and suppressions.xml into
new folder ip/config/checkstyle. Add some lines in build.gradle
to set up checkstyle.

Commit:
Merge branch 'branch-A-CheckStyle'

CaptainKeqing commented on src/main/java/harper/Duke.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * A chatbot called Harper.
+ */
+public class Duke {
```
Maybe it would make more sense to rename your class to Harper? Duke seems like a relic of the old version

CaptainKeqing commented on src/main/java/harper/tasks/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+    private LocalDateTime by;
+
+    /**
+     * Creates a task.Deadline task with description and deadline.
+     *
+     * @param description The description of the task.Deadline task.
+     * @param by The deadline of the task.
```
Should this javadoc  be cleaned up? Seems like a relic from intellij auto refactor.

I see this issue in other places

CaptainKeqing commented on src/main/java/harper/utils/Parser.java:
```diff
@@ -0,0 +1,174 @@
+            throw new HarperInvalidDeadlineException();
+        }
+        try {
+            LocalDateTime deadlineFormatted = LocalDateTime.parse(deadline, DATE_TIME_FORMATTER);
```
Should this variable be named more sensibly? something like formattedDeadline, similar to formattedEvent and formattedToDo

CaptainKeqing commented on src/main/java/harper/utils/Storage.java:
```diff
@@ -0,0 +1,107 @@
+     */
+    public void save(TaskList taskList) {
+        try {
+            FileWriter fw = new FileWriter(this.file);
```
Should this be named 'fileWriter' instead? Makes it easier to pronounce as mentioned [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/topics.html#w4-6-code-quality-naming)

CaptainKeqing reviewed (commented):
LGTM. Just a few nits in terms of naming.

I really like how you named your boolean variables though. Very clear to me

Commit:
Add GUI to the chatbot using javaFX

Commit:
Merge branch 'branch-Level-10'

Commit:
Add java assertions to verify program state and help debugging

No assertion is used.

It is hard to verify the assumptions made about the program state.
Assertions can help to verify them.

Assertions are added against null values to avoid NullPointerExceptions.
Also, assertions are added to verify commands passed to each command
handler are in correct length.

Let's use assertions to catch programmer's mistake and help debugging.

Commit:
Refactor code

Code quality can still be improved.

Code with good quality makes it more readable and understandable.

Removing ununsed class attribute and method. Packing some code fragment
into method to avoid multiple levels of abstraction. Declaring more
local boolean variables to improve readability.

This will not only help to improve readability, but also help for
debugging.

Let's refactor code to improve readability and helo debugging.

gosongying committed:
Merge pull request #1 from gosongying/branch-A-Assertions

Add java assertions to verify program state and help debugging

Commit:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/harper/utils/Parser.java
#	src/main/java/harper/utils/Storage.java

gosongying committed:
Merge pull request #2 from gosongying/branch-A-CodeQuality

Branch a code quality

Commit:
Implement update task feature

To update a task, it is required to delete it first then add another.

Task info might be updated frequently, by deleting it first then add is
not efficient.

Handling update command in Parser, declaring setter for each task type
to change fields and creating new exception class to handle invalid
update command.

Let's implement an update command to avoid using deleting which drops
efficiency.

Commit:
Merge branch 'branch-C-Update'

Commit:
Make some changes to method's header comment to follow the standard

Commit:
Change images displayed and add Ui.png

Commit:
Update README.md to add User Guide

Commit:
Build jar file
